File: backend/app/schemas/document_schema.py
from marshmallow import Schema, fields, validate


class SectionSchema(Schema):
    id = fields.Str(required=True)
    heading = fields.Str(required=True)
    content = fields.Str(required=True)
    order = fields.Int(required=True)


class DiagramSchema(Schema):
    diagram_type = fields.Str(required=True)
    mermaid_code = fields.Str(required=True)


class CodeFileSchema(Schema):
    filename = fields.Str(required=True)
    language = fields.Str(required=True)
    content_text = fields.Str(required=True)


class FigureSchema(Schema):
    fileUrl = fields.Str(required=False, allow_none=True)
    caption = fields.Str(required=False, allow_none=True)


class TemplateSettingsSchema(Schema):
    template = fields.Str(required=True, validate=validate.OneOf(["ieee", "college", "custom"]))
    darkPreview = fields.Bool(required=True)
    margins = fields.Dict(required=False)
    fontFamily = fields.Str(required=False)
    fontSize = fields.Int(required=False)


class DocumentSchema(Schema):
    meta = fields.Dict(required=True)
    sections = fields.List(fields.Nested(SectionSchema), required=True)
    diagrams = fields.List(fields.Nested(DiagramSchema), required=True)
    codeBlocks = fields.List(fields.Nested(CodeFileSchema), required=True)
    figures = fields.List(fields.Nested(FigureSchema), required=True)
    toc = fields.List(fields.Dict(), required=True)
    templateSettings = fields.Nested(TemplateSettingsSchema, required=True)

File: frontend/src/types/documentSchema.js
export const sectionKeys = [
  "introduction",
  "objective",
  "scope",
  "techStack",
  "modules",
  "conclusion",
];

export const defaultDocument = {
  meta: {
    projectTitle: "",
    studentName: "",
    guideName: "",
    professorName: "",
    techStack: "",
    year: "",
  },
  sections: sectionKeys.map((key, idx) => ({
    id: key,
    heading: `${idx + 1}. ${key.replace(/([A-Z])/g, " $1").replace(/^./, (c) => c.toUpperCase())}`,
    content: "",
    order: idx + 1,
  })),
  diagrams: [],
  codeBlocks: [],
  figures: [],
  toc: [],
  templateSettings: {
    template: "ieee",
    darkPreview: false,
    margins: { top: 64, right: 70, bottom: 64, left: 70 },
    fontFamily: "Source Serif 4",
    fontSize: 12,
  },
};
